contracts/cw721-stake/src/contract.rs

```diff
@@ -0,0 +1,1536 @@
+use crate::hooks::{stake_hook_msgs, unstake_hook_msgs};
+#[cfg(not(feature = "library"))]
+use crate::msg::{
+    ExecuteMsg, GetConfigResponse, GetHooksResponse, InstantiateMsg, Owner, QueryMsg,
+    StakedBalanceAtHeightResponse, TotalStakedAtHeightResponse,
+};
+use crate::state::{
+    Config, CONFIG, HOOKS, MAX_CLAIMS, NFT_CLAIMS, STAKED_NFTS_PER_OWNER, TOTAL_STAKED_NFTS,
+};
+use crate::ContractError;
+use cosmwasm_std::{
+    entry_point, to_binary, Binary, CosmosMsg, Deps, DepsMut, Empty, Env, MessageInfo, Response,
+    StdError, StdResult, Uint128, WasmMsg,
+};
+use cw2::set_contract_version;
+use cw721::Cw721ReceiveMsg;
+use cw721_controllers::NftClaimsResponse;
+use cw_utils::Duration;
+use std::collections::HashSet;
+use std::convert::{From, TryFrom};
+
+const CONTRACT_NAME: &str = "crates.io:stake_cw721";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response<Empty>, ContractError> {
+    let owner = msg
+        .owner
+        .as_ref()
+        .map(|owner| match owner {
+            Owner::Addr(address) => deps.api.addr_validate(address),
+            Owner::Instantiator {} => Ok(info.sender),
+        })
+        .transpose()?;
+    let manager = msg
+        .manager
+        .as_ref()
+        .map(|h| deps.api.addr_validate(h))
+        .transpose()?;
+
+    let config = Config {
+        owner: owner.clone(),
+        manager,
+        nft_address: deps.api.addr_validate(&msg.nft_address)?,
+        unstaking_duration: msg.unstaking_duration,
+    };
+    CONFIG.save(deps.storage, &config)?;
+    TOTAL_STAKED_NFTS.save(deps.storage, &Uint128::zero(), env.block.height)?;
```

Comment:
nit: is this strictly necessary, as long as we use `may_load_at_height`

Reply:
I've been reflecting on this a little bit and I'm leaning towards the initialize to zero and unwrap approach. My thought is that this more clearly delineates when the contract is behaving as expected. 

For example: say an attacker is somehow able to remove the `TOTAL_STAKED_NFTS` storage data - if we use `unwrap_or_default` this is invisible as we'll just start back at zero. If we initialize to zero and then unwrap post-attack the operations will fail and the data invalidation will be caught.

What do you think about this? My thinking here is dependent on the idea what we'd rather have methods fail than unwrap and continue forward.